src/Kiota.Builder/Writers/Go/CodeMethodWriter.cs

```diff
@@ -553,30 +554,35 @@ private void WriteDeserializerBodyForIntersectionModel(CodeMethod method, CodeCl
-            writer.WriteLine($"res[\"{property.SerializationName ?? property.Name.ToFirstCharacterLowerCase()}\"] = func (n {parsableImportSymbol}) error {{");
-            writer.IncreaseIndent();
-            var propertyTypeImportName = conventions.GetTypeString(property.Type, parentClass, false, false);
-            var deserializationMethodName = GetDeserializationMethodName(property.Type, parentClass);
-            writer.WriteLine($"val, err := n.{deserializationMethodName}");
-            WriteReturnError(writer);
-            writer.WriteLine("if val != nil {");
-            writer.IncreaseIndent();
-            var (valueArgument, pointerSymbol, dereference) = (property.Type.AllTypes.First().TypeDefinition, property.Type.IsCollection) switch {
-                (CodeClass, false) or (CodeEnum, false) => (GetTypeAssertion("val", $"*{propertyTypeImportName}"), string.Empty, true),
-                (CodeClass, true) or (CodeEnum, true) => ("res", "*", true),
-                (CodeInterface, false) => (GetTypeAssertion("val", propertyTypeImportName), string.Empty, false),
-                (CodeInterface, true) => ("res", string.Empty, false),
-                (_, true) => ("res", "*", true),
-                _ => ("val", string.Empty, true),
+
+        private void WriteFieldDeserializer(CodeProperty property, LanguageWriter writer, CodeClass parentClass,string parsableImportSymbol)
+        {
+            var (_, sourceArgument) = GetDeserializationMethodNameAndFactory(property.Type, parentClass);
+
+            var utilitySetter = (property.Type.AllTypes.First().TypeDefinition, property.Type.IsCollection) switch
+            {
+                (CodeClass, false) => "SetObjectValue",
+                (CodeClass, true) => "SetCollectionOfObjectValues",
+                (CodeEnum, false) => "SetEnumValue",
+                (CodeEnum, true) => "SetCollectionOfEnumValues",
+                (CodeInterface, false) => "SetObjectValue",
+                (CodeInterface, true) => "SetCollectionOfObjectValues",
+                (_, true) => "SetCollectionOfPrimitiveValues",
+                _ => $"Set{getSerializerTypeName(property)}Value",
             };
-            if(property.Type.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None)
-                WriteCollectionCast(propertyTypeImportName, "val", "res", writer, pointerSymbol, dereference);
-            writer.WriteLine($"m.{property.Setter.Name.ToFirstCharacterUpperCase()}({valueArgument})");
-            writer.CloseBlock();
-            writer.WriteLine("return nil");
-            writer.CloseBlock();
+
+            var methodFactory = sourceArgument == String.Empty ? "" : $"{sourceArgument} , ";
+            writer.WriteLine($"res[\"{property.SerializationName ?? property.Name.ToFirstCharacterLowerCase()}\"] = utils.{utilitySetter}({methodFactory}m.{property.Setter.Name.ToFirstCharacterUpperCase()})");
```

Comment:
maybe move the utils package name to the convention service?

Reply:
After further reflections, this should in facts be the abstractions hash

Reply:
done